core/src/stored/backends/dplcompat_device.cc

```diff
@@ -0,0 +1,418 @@
+    Dmsg1(debug_info, "Not flushing empty chunk %s/%s\n", obj_name.c_str(),
+          obj_chunk.c_str());
+    return true;
+  }
+  Dmsg1(debug_trace, "Flushing chunk %s/%s\n", obj_name.c_str(),
+        obj_chunk.c_str());
+
+  auto inflight_lease = getInflightLease(request);
+  if (!inflight_lease) {
+    Dmsg0(debug_info, "Could not acquire inflight lease for %s %s\n",
+          obj_name.c_str(), obj_chunk.c_str());
+    return false;
+  }
+
+
+  /* Check on the remote backing store if the chunk already exists.
+   * We only upload this chunk if it is bigger then the chunk that exists
+   * on the remote backing store. When using io-threads it could happen
+   * that there are multiple flush requests for the same chunk when a
+   * chunk is reused in a next backup job. We only want the chunk with
+   * the biggest amount of valid data to persist as we only append to
+   * chunks. */
+  auto obj_stat = m_storage.stat(obj_name, obj_chunk);
+
+  if (obj_stat && obj_stat->size > request->wbuflen) {
+    Dmsg1(debug_info,
+          "Not uploading chunk %s with size %d, as chunk with size %d is "
+          "already present\n",
+          obj_name.c_str(), obj_stat->size, request->wbuflen);
+    return true;
+  }
+
+  auto obj_data = gsl::span{request->buffer, request->wbuflen};
+  Dmsg1(debug_info, "Uploading %zu bytes of data\n", request->wbuflen);
+  if (auto result = m_storage.upload(obj_name, obj_chunk, obj_data)) {
+    return true;
+  } else {
+    PmStrcpy(errmsg, result.error().c_str());
+    dev_errno = EIO;
+    return false;
+  }
+}
+
+// Internal method for reading a chunk from the remote backing store.
+bool DropletCompatibleDevice::ReadRemoteChunk(chunk_io_request* request)
+{
+  const std::string obj_name = request->volname;
+  const std::string obj_chunk = get_chunk_name(request);
+  Dmsg1(debug_trace, "Reading chunk %s\n", obj_name.c_str());
+
+  // check object metadata
+  auto obj_stat = m_storage.stat(obj_name, obj_chunk);
+  if (!obj_stat) {
+    PmStrcpy(errmsg, obj_stat.error().c_str());
+    Dmsg1(debug_info, "%s", errmsg);
+    dev_errno = EIO;
+    return false;
+  } else if (obj_stat->size > request->wbuflen) {
+    Mmsg3(errmsg,
+          T_("Failed to read %s (%ld) to big to fit in chunksize of %ld "
+             "bytes\n"),
+          obj_name.c_str(), obj_stat->size, request->wbuflen);
+    Dmsg1(debug_info, "%s", errmsg);
+    dev_errno = EINVAL;
+    return false;
+  }
+
+  if (auto obj_data = m_storage.download(obj_name, obj_chunk,
+                                         {request->buffer, obj_stat->size})) {
+    *request->rbuflen = obj_data->size_bytes();
+    return true;
+  } else {
+    PmStrcpy(errmsg, obj_data.error().c_str());
+    Dmsg1(debug_info, "%s", errmsg);
+    dev_errno = EIO;
+    return false;
+  }
+}
+
+bool DropletCompatibleDevice::TruncateRemoteVolume(DeviceControlRecord*)
+{
+  const char* vol_name = getVolCatName();
+  const auto list = m_storage.list(vol_name);
```

Comment:
I think `chunks` is a better name.
```suggestion
  const auto chunks = m_storage.list(vol_name);
```